### DISCUSSION TITLE
[Backend] 성공/실패시 기본 응답 형태 정의

### DISCUSSION THREAD

**Discussion by rkskekzzz** [80000Coding/80000Coding-Web-Client#61] [General]
# 현재 상황
현재 다양한 팀으로 나누어서 기능구현을 하고 있는데 API 요청 성공/실패시의 기본 응답 형태 정의가 필요해 보입니다!

아래는 간단한 예시를 만들어 보았는데 이렇게 메세지, data 등을 동일한 포맷으로 보내면 좋을 것 같습니다!
원하는 방식이 있으면 편하게 말씀해주세요!!

```javascript
// 성공
{
  message : "success",
  data : { ... }
}
// 실패
{
  message : "fail",
  data : { ... }
}
```

**Comment by ChanHoHan** [comment]
code, message, data, etc 이정도에 필요하면 link정도? 추가하면 어떨까요?
추가적으로, 페이지를 조회한다면 그때만  totalPages, totalElements 정도가 들어가면 될거같습니다

ex>
{
  "code": "200_00", //  required > 또는, 20000 ? 이건 정하기 나름일거 같습니다
  "meesage": "success", // required이면 보낼 값 없으면 null or ""
  "data": { // required이면 보낼 값 없으면 null or ""
       "contentId": 12,
         ...
    },
   "etc": "추가적인 정보", // optional
   "totalPages": 10, // optional > 전체 페이지 수
   "totalElements": 100 // optional > 전체 갯수
}

추상 클래스를 만들고, 모든 컨트롤러에서 그 클래스를 상속받아서 구현하는 식으로 하면, 모든 API마다 동일한 구조로 Response를 내릴수 있을거 같습니다.

더 좋은 의견있으면 말씀 부탁드릴게요

**Reply by rkskekzzz** [reply]
여기서 code는 어떤 정보를 담고 있는지 궁금합니다!!

**Reply by ChanHoHan** [reply]
예를들어, HTTP STATUS가 400인 경우에도 여러가지 케이스가 있잖아요? (요청인자가 올바르지 않다든지, 아이디, 비밀번호를 잘못 입력했다든지 등등)
그걸 메세지로만 판단하기보다는, 코드로 관리하면 프론트엔드에선 그 코드값을 catch로 잡아서 다음 진행사항을 결정할 수 있지 않을까 했습니다.
대신.. 코드를 문서화 해야겠죠 ㅠ

**Reply by psychology50** [reply]
저 개인적으로 Response의 body에 code 정보를 넣는 건 별로 안 좋다고 생각해요.  
예전에 모 개발 팀장 분하고도 이거 관련으로 대화해봤었는데, status를 관리하는 방법이 Response 객체에 따로 있던 걸로 알고 있습니다.
[참고 링크](https://velog.io/@jakeseo_me/Spring-Boot%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%9C-RESTful-Web-Services-%EA%B0%9C%EB%B0%9C-14-HTTP-Status-Code-%EC%A0%9C%EC%96%B4)  
아마 찬호님께서 말씀하신 게 이 내용일 거라 생각해요

**Reply by ChanHoHan** [reply]
앗 보내주신 링크는 https://developer.mozilla.org/en-US/docs/Web/HTTP/Status 에 정의된 상태코드를 말씀하시는거 같구요
![image](https://github.com/80000Coding/80000Coding-Web-Client/assets/46598292/493ad7ca-0026-4d0c-a395-2acacbca67d2)
[상태코드]

제가 말씀드린건, 저희가 response body에 json을 보낼텐데, 
저희 내부적으로 code라는 규칙을 정해서, 프로튼엔드에게 API 요청에 대한 응답을 정의한 code로 내려주면 된다고 생각한겁니다 ㅎㅎ

요청을 보냈는데, 프론트엔드에서 http status code 400으로만 분기처리를 할수밖에 없다면, 정확히 어떤 오류인지 알지 못하기 때문에 그 후속 작업을 하기가 제한적이잖아요? 그래서 400_01 (요청 인자의 오류가 발생) 이런 code를 보내서, 프론트엔드에서는 fetch 요청으로 보냈을때 에러가 나면 어떻게 분기처리를 할지 정할 수 있을 거라 생각했습니다.

**Reply by codero225** [reply]
오류가 여러개 날 수 있으니 해당 내용에 따라 세부적으로 표현하자는 이야기인거죠?

**Reply by ChanHoHan** [reply]
넵 그렇습니다!

**Reply by psychology50** [reply]
아 ㅋㅋ 이제 누가 code 문서화 해줄지 폭탄 돌리기 시작

**Reply by codero225** [reply]
기능 만든 사람이 책임지고 swagger에 작성하기? ㅎㅎ

**Reply by psychology50** [reply]
![image](https://github.com/80000Coding/80000Coding-Web-Client/assets/96044622/df27578c-0a6e-4fe4-af52-7d83b61e0bf3)

**Comment by psychology50** [comment]
어제부터 곰곰이 생각은 해봤는데, 진짜 뭐가 좋을 지 잘 모르겠습니다.  
message가 들어간다면 굳이 뭘 써야 할까? 싶기만 해서..

**Reply by rkskekzzz** [reply]
message를 꼭 넣을 필요는 없습니다!

**Reply by rkskekzzz** [reply]
아래는 토스페이먼츠 SDK에서 요청 실패 응답 예시인데요

이런 방식으로 에러 응답형태가 통일만 되면 좋을 것 같습니다 ㅎㅎ
![image](https://github.com/80000Coding/80000Coding-Web-Client/assets/57925497/70302289-493d-4aac-8918-da3d48b862df)

**Reply by ChanHoHan** [reply]
message는 프론트엔드 분들께서 작업할 때, 정확히 어떤 에러가 발생한지 알기 힘든 경우를 대비해서 넣을려고 했습니다.
꼭 안넣어도 된다고 하면, 저희도 안넣는걸로 통일하면 될거같아요 ㅎㅎ

**Reply by rkskekzzz** [reply]
찬호님께서 말씀하신 것 처럼 프론트에서 사용자에게도 해당 내용을 전달할 때 쓸 수 있긴합니다!! 

만약 같은 상태코드에서 에러가 세분화 된다면 message를 추가해 주시면 좋을 것 같습니다 ㅎㅎ

**Reply by ChanHoHan** [reply]
저도 회사에서 이런 고민을 했었던 적이 있는데요.
프론트엔드에서 해당 message를 그대로 화면에 노출할 경우에, 만약 그 문구를 수정해야할 사항이 올 경우에 백엔드를 수정해야한다는 단점이 있더라구요.
그래서 message를 사용하는건 상관없지만, 웬만하면 code나 http status code를 가지고 직접 프론트엔드에서 관리하는게 좋을거 같습니다 ㅎㅎ

**Reply by rkskekzzz** [reply]
토스 예시처럼 코드로 보내주는 방법도 저는 좋아보입니다! 말씀해주신대로 message는 디버깅용으로 사용하구요!(만약 만들면..?)

**Reply by psychology50** [reply]
실제로 저런 식으로 하는 군요. 처음 알았습니다 허허

**Reply by ChanHoHan** [reply]
그럼 code, message, data는 필수값 확정으로 할까요?!

**Reply by rkskekzzz** [reply]
@ChanHoHan 실패 성공 모두 다 포함되는건가요?

**Reply by ChanHoHan** [reply]
@rkskekzzz 네네! 보낼거 없으면, null로 보낼지 아님 빈문자열로 보낼지만 정하면 될거같습니다

**Reply by psychology50** [reply]
답이 늦어서 죄송합니다! 관련해서 이것저것 찾아보고 직접 테스트해볼 것들이 있어서 많이 지연됐어요.
  
디코에 올린 내용이긴 한데, [페이스북에서는 이렇게 에러 코드를 관리](https://developers.facebook.com/docs/marketing-api/error-reference?locale=ko_KR)  한다고 하더라구요.  
  
제가 생각해봤을 때는 아래처럼 포맷이 정해지면 좋을 거 같아요.  
  
1. 성공했을 때 
```json
{
    "count": 1,
    "data": {
         
    }
}
```
  
2. 실패했을 때
```json
{
    "code": -101,
    "msg": "NotRegisteredUserException"
}
```
  
에러 코드의 경우에 [해당 블로그](https://adjh54.tistory.com/79#1.%20%EC%84%B1%EA%B3%B5%20%EC%9D%91%EB%8B%B5%20%EC%BD%94%EB%93%9C%20%EA%B5%AC%EC%84%B1-1)처럼 작성해보는 것도 재밌을 것 같아요.  
```java
/**
 * [공통 코드] API 통신에 대한 '에러 코드'를 Enum 형태로 관리를 한다.
 * Global Error CodeList : 전역으로 발생하는 에러코드를 관리한다.
 * Custom Error CodeList : 업무 페이지에서 발생하는 에러코드를 관리한다
 * Error Code Constructor : 에러코드를 직접적으로 사용하기 위한 생성자를 구성한다.
 *
 * @author lee
 */
@Getter
public enum ErrorCode {

    /**
     * ******************************* Global Error CodeList ***************************************
     * HTTP Status Code
     * 400 : Bad Request
     * 401 : Unauthorized
     * 403 : Forbidden
     * 404 : Not Found
     * 500 : Internal Server Error
     * *********************************************************************************************
     */
    // 잘못된 서버 요청
    BAD_REQUEST_ERROR(400, "G001", "Bad Request Exception"),

    // @RequestBody 데이터 미 존재
    REQUEST_BODY_MISSING_ERROR(400, "G002", "Required request body is missing"),

    // 유효하지 않은 타입
    INVALID_TYPE_VALUE(400, "G003", " Invalid Type Value"),

    // Request Parameter 로 데이터가 전달되지 않을 경우
    MISSING_REQUEST_PARAMETER_ERROR(400, "G004", "Missing Servlet RequestParameter Exception"),

    // 입력/출력 값이 유효하지 않음
    IO_ERROR(400, "G005", "I/O Exception"),

    // com.google.gson JSON 파싱 실패
    JSON_PARSE_ERROR(400, "G006", "JsonParseException"),

    // com.fasterxml.jackson.core Processing Error
    JACKSON_PROCESS_ERROR(400, "G007", "com.fasterxml.jackson.core Exception"),

    // 권한이 없음
    FORBIDDEN_ERROR(403, "G008", "Forbidden Exception"),

    // 서버로 요청한 리소스가 존재하지 않음
    NOT_FOUND_ERROR(404, "G009", "Not Found Exception"),

    // NULL Point Exception 발생
    NULL_POINT_ERROR(404, "G010", "Null Point Exception"),

    // @RequestBody 및 @RequestParam, @PathVariable 값이 유효하지 않음
    NOT_VALID_ERROR(404, "G011", "handle Validation Exception"),

    // @RequestBody 및 @RequestParam, @PathVariable 값이 유효하지 않음
    NOT_VALID_HEADER_ERROR(404, "G012", "Header에 데이터가 존재하지 않는 경우 "),

    // 서버가 처리 할 방법을 모르는 경우 발생
    INTERNAL_SERVER_ERROR(500, "G999", "Internal Server Error Exception"),


    /**
     * ******************************* Custom Error CodeList ***************************************
     */
    // Transaction Insert Error
    INSERT_ERROR(200, "9999", "Insert Transaction Error Exception"),

    // Transaction Update Error
    UPDATE_ERROR(200, "9999", "Update Transaction Error Exception"),

    // Transaction Delete Error
    DELETE_ERROR(200, "9999", "Delete Transaction Error Exception"),

    ; // End

    /**
     * ******************************* Error Code Constructor ***************************************
     */
    // 에러 코드의 '코드 상태'을 반환한다.
    private final int status;

    // 에러 코드의 '코드간 구분 값'을 반환한다.
    private final String divisionCode;

    // 에러 코드의 '코드 메시지'을 반환한다.
    private final String message;

    // 생성자 구성
    ErrorCode(final int status, final String divisionCode, final String message) {
        this.status = status;
        this.divisionCode = divisionCode;
        this.message = message;
    }
}
```

**Reply by psychology50** [reply]
아무리 생각해도 message에 success와 fail을 담는 건 낭비라고 생각했습니다.  
그래서 에러를 확실하게 분리하되, 가끔 status 200 OK가 뜨면서 "fail"로 뜨는 경우가 있습니다.  
(예를 들어, 회원 데이터를 조회했는데 없는 경우. packet은 정상적으로 오고 갔으나 예외 처리로 인해 message에 "fail"이 뜹니다.)

이런 경우를 대비해서, 마치 비검사 예외처리 방식처럼 "값이 없음"을 count로 알려주면 클라이언트에서 처리할 수 있다고 생각합니다.  
동시에 data의 사이즈를 알 수 있으므로 "fail"을 보내는 것보다 훨씬 유용한 정보로 활용 가능하구요.

**Reply by rkskekzzz** [reply]
좋아요 ㅎㅎ 에러 코드는 백엔드 분들이 상황에 맞게 만들어주시면 참고하겠습니당!

**Reply by ChanHoHan** [reply]
@rkskekzzz 성공, 실패의 JSON 필드가 달라도 상관 없는거죠 ??

**Reply by rkskekzzz** [reply]
넵 상관없습니다!

**Reply by ChanHoHan** [reply]
@psychology50
개인적으로 시퀀스 코드 처럼 보이는 페이스북 방식보다는,
두번째로 올려주신 division code처럼 Global Error인 경우, 앞에 G를 붙이는 식이 좋은거 같네요. 재밌을거 같아요 !

**Reply by ChanHoHan** [reply]
@psychology50 성공시에 count가 data의 사이즈를 의미하는건가요?

**Reply by psychology50** [reply]
넵, 맞습니다.  
말씀드렸듯이 데이터가 없어서 조회에 실패하는 경우가 네트워크 문제와는 관련이 없기 때문에,  
해당 케이스를 처리하기 위해서 count가 0임을 알려주면 프론트 단에서 조회에 실패했다고 파악할 수 있을 거라 생각했습니다.  
그 외에도 data size를 알면 유용하게 사용할 수 있을 거라 생각해서 count를 넣는 걸 고려해봤어요.

**Comment by dnjsals45** [comment]
블로그글과 같이 에러 코드를 작성하게 된다면, 모든 에러 상황을 다같이 고민해서 만든 후에 그 코드들만 이용하는 방식이 될까요?

**Reply by ChanHoHan** [reply]
우선, 실시간으로 공유할 수 있는 에러 노트 페이지를 만드는게 좋을거 같구요. (notion)
글로벌 에러는 다같이 만들고, 
에러를 추가해야한다 싶으면, 에러 노트 페이지에 추가 후, 
본인의 feature에서 코드를 추가한 다음 바로 사용하는게 어떨까 싶습니다.

**Reply by rkskekzzz** [reply]
노션에 에러 관리를 위한 페이지를 만들어 두곘습니다 :)
